sentry-core/src/main/java/io/sentry/core/SentryClient.java

```diff
@@ -160,28 +203,22 @@ void updateSessionData(
                 }
 
                 if (session.update(status, userAgent, crashedOrErrored)) {
-
-                  Object sessionHint;
-
                   // if hint is DiskFlushNotification, it means we have an uncaughtException
                   // and we can end the session.
                   if (hint instanceof DiskFlushNotification) {
-                    sessionHint = new SessionEndHint();
                     session.end();
-                  } else {
-                    // otherwise we just cache in the disk but do not flush to the network.
-                    sessionHint = new SessionUpdateHint();
                   }
-                  captureSession(session, sessionHint);
                 }
               } else {
                 options.getLogger().log(SentryLevel.INFO, "Session is null on scope.withSession");
               }
             });
+        clonedSession = scope.cloneSession();
```

Comment:
We're cloning outside of the lock so we could be cloning a different session than the one updated above. For example if a thread closing a session was waiting on a lock to close the session and takes the lock before we call `clone` here.

Maybe we clone it inside the callback and assign the reference to a variable declared outside of the block. Ugly but this is guaranteed to clone the right state.

The method on the session (like `update`) could return the clone on each operation already to avoid having that public clone that also takes a lock

Reply:
good catch, I'll need to use an `AtomicReference` maybe because it's a lambda and I cannot assign the value to not even a final object outside of the lambda or to refactor a little bit this, i will have a look